frontend/src/upload.js

```diff
@@ -26,6 +26,34 @@ if (storage.has('referrer')) {
 $(document).ready(function() {
   $('#file-upload').change(onUpload);
 
+  if ('serviceWorker' in navigator && 'PushManager' in window) {
+    navigator.serviceWorker.register('serviceWorker.js')
+    .then(function(registration) {
+      return registration.pushManager.getSubscription()
+      .then(function(subscription) {
+        if (subscription) {
+          return subscription;
+        }
+        return registration.pushManager.subscribe({ userVisibleOnly: true });
+      });
+    }).then(function(subscription) {
+      const rawKey = subscription.getKey ? subscription.getKey('p256dh') : '';
+      const key = rawKey ?
+        btoa(String.fromCharCode.apply(null, new Uint8Array(rawKey))) : '';
```

Comment:
Only duplicated twice, but not sure if you want to wrap this in a helper function.